apollo-federation/src/sources/connect/validation/selection/visitor.rs

```diff
@@ -0,0 +1,288 @@
+use crate::sources::connect::expand::visitors::FieldVisitor;
+use crate::sources::connect::expand::visitors::GroupVisitor;
+use crate::sources::connect::json_selection::LitExpr;
+use crate::sources::connect::json_selection::MethodArgs;
+use crate::sources::connect::json_selection::NamedSelection;
+use crate::sources::connect::json_selection::PathList;
+use crate::sources::connect::JSONSelection;
+use crate::sources::connect::SubSelection;
+
+/// A JSON Selection visitor
+pub(super) trait Visitor {
+    type Error;
+
+    /// Visit a JSON Selection part. The visitor is guaranteed that all children of this part will
+    /// be visited before [`Visitor::end_visit`] is called. A mutable reference is required since
+    /// the visitor will typically require modifying internal state.
+    fn visit(&mut self, part: &SelectionPart) -> Result<(), Self::Error>;
+
+    /// End visiting a JSON Selection part. This is called after all children of a part have been
+    /// visited. The default implementation does nothing.
+    fn end_visit(&mut self, _part: &SelectionPart) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+/// A part of a JSON Selection to be visited
+#[derive(Clone, Debug)]
+pub(super) enum SelectionPart<'schema> {
+    JSONSelection(&'schema JSONSelection),
+    LitExpr(&'schema LitExpr),
+    MethodArgs(&'schema MethodArgs),
+    NamedSelection(&'schema NamedSelection),
+    PathList(&'schema PathList),
+    SubSelection(&'schema SubSelection),
+}
+
+/// Visit a JSON Selection, invoking a visitor on each part
+pub(super) fn visit<V: Visitor>(selection: &JSONSelection, mut visitor: V) -> Result<(), V::Error> {
+    SelectionWalker {
+        visitor: &mut visitor,
+        stack: vec![],
+    }
+    .walk(SelectionGroup::Root {
+        children: vec![SelectionPart::JSONSelection(selection)],
+    })
+}
+
+/// Walks a JSON Selection and invokes a visitor
+struct SelectionWalker<'schema, V: Visitor> {
+    pub(super) visitor: &'schema mut V,
+    pub(super) stack: Vec<SelectionPart<'schema>>,
+}
+
+enum SelectionGroup<'schema> {
+    Root {
+        children: Vec<SelectionPart<'schema>>,
+    },
+    Child {
+        parent: SelectionPart<'schema>,
+        children: Vec<SelectionPart<'schema>>,
+    },
+    Empty {
+        parent: SelectionPart<'schema>,
+    },
+}
```

Comment:
Could this be simplified to avoid the `Root`/`Child` distinction by having a single `parent: Option<SelectionPart<'schema>>` that would be `None` in the `Root` case?

It also seems like `Empty` is just the case where `children` is empty. If that's true, I don't think this type benefits much from being an enum with three different cases, when it could be a single struct.

Reply:
The `Empty` one was just vestigial and related to something I abandoned - I'll remove that. The `Root` vs `Child` is a way to use Rust type system to ensure correctness of data. In other languages, you might have a parent that is `Optional` or can be `null`. This allows creation of instances that are incorrect. For example, you could have a node return a child that has its parent set to `Empty` or `null`, which would be invalid. Every bit of code that accesses the parent field then needs to check for this and somehow handle it to avoid hitting a `NullPointerExdeption` or the like - but there isn't really anything sensible they can do if they encounter it. In Rust, we can avoid this situation entirely by easily creating type variants that enforce correctness of data. A `Root` is the only node that has no parent, and everything else requires a parent.